src/cgutils.cpp

```diff
@@ -2120,12 +2119,15 @@ static jl_cgval_t typed_store(jl_codectx_t &ctx,
         FailOrder = AtomicOrdering::Monotonic;
     unsigned nb = isboxed ? sizeof(void*) : jl_datatype_size(jltype);
     AllocaInst *intcast = nullptr;
+    Type *intcast_eltyp = nullptr;
+    bool tracked_pointers = isboxed || CountTrackedPointers(elty).count > 0;
     if (!isboxed && Order != AtomicOrdering::NotAtomic && !elty->isIntOrPtrTy()) {
+        intcast_eltyp = elty;
+        elty = Type::getIntNTy(ctx.builder.getContext(), 8 * nb);
```

Comment:
Does this need a check so that N isn't too large? Or does the code that falls back to a lock avoid that?

Reply:
there is no limit on how large N can be per the standard (but yes, we limit to MAX_ATOMIC_SIZE because it generates better code)